src/lib/API/Repository/Value/Filter/Filter.php

```diff
@@ -0,0 +1,155 @@
+     *
+     * @see \eZ\Publish\API\Repository\ContentService::createFindFilter()
+     * @see \eZ\Publish\API\Repository\LocationService::createFindFilter()
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\BadStateException for invalid Sort Clause
+     */
+    public function __construct(?Criterion $criterion = null, array $sortClauses = [])
+    {
+        $this->criterion = $criterion;
+        foreach ($sortClauses as $idx => $sortClause) {
+            if (!$sortClause instanceof SortClause) {
+                throw new BadStateException(
+                    '$sortClauses',
+                    sprintf(
+                        'Expected an instance of "%s", got "%s" at position %d',
+                        SortClause::class,
+                        is_object($sortClause) ? get_class($sortClause) : gettype($sortClause),
+                        $idx
+                    )
+                );
+            }
+
+            $this->sortClauses[] = $sortClause;
+        }
+    }
+
+    /**
+     * Reset Filter so it can be built from scratch.
+     */
+    public function reset(): Filter
+    {
+        $this->criterion = null;
+        $this->sortClauses = [];
+
+        return $this;
+    }
+
+    /**
+     * Set filtering Criterion.
+     *
+     * If multiple Criteria are required, either use `andWithCriterion`/`orWithCriterion` or wrap
+     * them with Logical operator Criterion.
+     *
+     * To re-build Criterion from scratch `reset` it first.
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\BadStateException if Criterion is already set
+     *
+     * @see reset
+     * @see andWithCriterion
+     * @see orWithCriterion
+     * @see \eZ\Publish\API\Repository\Values\URL\Query\Criterion\LogicalOr
+     * @see \eZ\Publish\API\Repository\Values\URL\Query\Criterion\LogicalAnd
+     */
+    public function withCriterion(Criterion $criterion): Filter
+    {
+        if (null !== $this->criterion) {
+            throw new BadStateException(
+                '$criterion',
+                'Criterion is already set. ' .
+                'To append Criterion invoke either "andWithCriterion" or "orWithCriterion". ' .
+                'To start building Criterion from scratch "reset" it first.'
+            );
+        }
+
+        $this->criterion = $criterion;
+
+        return $this;
+    }
+
+    /**
+     * @see withCriterion
+     */
+    public function andWithCriterion(Criterion $criterion): Filter
+    {
+        if (null === $this->criterion) {
+            // for better DX allow operation on uninitialized Criterion by setting it as-is
+            $this->criterion = $criterion;
+        } elseif ($this->criterion instanceof Criterion\LogicalAnd) {
+            $this->criterion->criteria[] = $criterion;
+        } else {
+            $this->criterion = new Criterion\LogicalAnd([$this->criterion, $criterion]);
+        }
+
+        return $this;
+    }
```

Comment:
Q: Could we split this class into `Filter` and and `FilterBuilder`? 

* `Filter` will be VO (inc. immutable)
* `FilterBuilder` could be an interface and concrete implementation could be hidden in Core namespace
* Better responsibilities separation

Reply:
> Q: Could we split this class into `Filter` and and `FilterBuilder`?
> 
> * `Filter` will be VO (inc. immutable)
> * `FilterBuilder` could be an interface and concrete implementation could be hidden in Core namespace
> * Better responsibilities separation

I don't mind, WDYT @bdunogier? :)

Reply:
Hmm, it's not a builder, it's an actual value object. The way I suggested it, `with*` methods are fluent setters, because they can manipulate an immutable of their type (clone self, modify the clone, and return the clone).

There isn't enough flesh for a builder imho. A "real" builder would be "dynamic", like the one I had shown a while ago.